include/irr/asset/ITexturePacker.h

```diff
@@ -0,0 +1,483 @@
+    {
+        //free physical pages
+        auto extent = _img->getCreationParameters().extent;
+        const uint32_t levelCount = countLevelsTakingAtLeastOnePage(extent, _subres);
+
+        for (uint32_t i = 0u; i < levelCount; ++i)
+        {
+            const uint32_t w = (std::max(extent.width>>(_subres.baseMipLevel+i),1u) + m_pgSzxy-1u) / m_pgSzxy;
+            const uint32_t h = (std::max(extent.height>>(_subres.baseMipLevel+i),1u) + m_pgSzxy-1u) / m_pgSzxy;
+
+            uint32_t* const pgTab = reinterpret_cast<uint32_t*>(
+                reinterpret_cast<uint8_t*>(m_pageTable->getBuffer()->getPointer()) + m_pageTable->getRegions().begin()[i].bufferOffset
+                );
+            const uint32_t pgtPitch = m_pageTable->getRegions().begin()[i].bufferRowLength;
+            const uint32_t offset = (_addr.y>>i)*pgtPitch + (_addr.x>>i);
+
+            for (uint32_t y = 0u; y < h; ++y)
+                for (uint32_t x = 0u; x < w; ++x)
+                {
+                    const uint32_t linTexelAddr = offset + y*pgtPitch + x;
+
+                    SPhysPgOffset physPgOffset = pgTab[linTexelAddr];
+                    if (physPgOffset_valid(physPgOffset))
+                    {
+                        pgTab[linTexelAddr] = SPhysPgOffset::invalid_addr;
+
+                        uint32_t addrs[2] { physPgOffset.addr&0xffffu, physPgOffset_mipTailAddr(physPgOffset).addr&0xffffu };
+                        const uint32_t szs[2]{ 1u,1u };
+                        core::address_allocator_traits<phys_pg_addr_alctr_t>::multi_free_addr(m_physPgAddrAlctr, physPgOffset_hasMipTailAddr(physPgOffset) ? 2u : 1u, addrs, szs);
+                    }
+                }
+        }
+
+        //free entries in page table
+        ITexturePacker::free(_addr, _img, _subres);
+    }
+
+    page_tab_offset_t pack(const ICPUImage* _img, const ICPUImage::SSubresourceRange& _subres)
+    {
+        if (getFormatClass(_img->getCreationParameters().format)!=getFormatClass(m_physAddrTex->getCreationParameters().format))
+            return page_tab_offset_invalid();
+
+        const auto extent = _img->getCreationParameters().extent;
+
+        if ((extent.width>>_subres.baseMipLevel) > maxAllocatableTextureSz() || (extent.height>>_subres.baseMipLevel) > maxAllocatableTextureSz())
+            return page_tab_offset_invalid();
+
+        const page_tab_offset_t pgtOffset = alloc(_img, _subres);
+        if ((pgtOffset==page_tab_offset_invalid()).all())
+            return pgtOffset;
+
+        const uint32_t levelCount = countLevelsTakingAtLeastOnePage(extent, _subres);
+
+        uint32_t miptailPgAddr = SPhysPgOffset::invalid_addr;
+
+        const uint32_t texelSz = getTexelOrBlockBytesize(m_physAddrTex->getCreationParameters().format);
+        //fill page table and pack present mips into physical addr texture
+        for (uint32_t i = 0u; i < _subres.levelCount; ++i)
```

Comment:
this needs to be a minimum, `core::min(_subres.levelCount,m_maxPageTableMipLevels+1u)`

Reply:
leaving the loop unreviewed for now.